File: Jobs2/without_latent/py/scenarios.py
# =============================================================================
# scenarios.py — JOBS II semi-synthetic, 8 scenarios per Cheng et al. (2022).
# Mirror of R/scenarios.R — order must match exactly.
# =============================================================================

import os
import itertools

# expand.grid in R varies first column FASTEST, so we replicate that order:
#   n varies fastest, then eta, then alpha_share
_N_LIST           = [500, 1000]
_ETA_LIST         = [1, 10]
_ALPHA_SHARE_LIST = [0.10, 0.50]

SCENARIOS = []
for share in _ALPHA_SHARE_LIST:
    for eta in _ETA_LIST:
        for n in _N_LIST:
            SCENARIOS.append({
                "n":           n,
                "eta":         eta,
                "alpha_share": share,
                "name":        f"n{n}_eta{eta}_share{int(share * 100)}",
            })

for i, s in enumerate(SCENARIOS):
    s["scenario_id"] = i + 1

N_REPS_PER_SCENARIO = 100
N_SCENARIOS         = len(SCENARIOS)
N_TASKS             = N_SCENARIOS * N_REPS_PER_SCENARIO


def task_id_to_indices(task_id: int) -> dict:
    if task_id < 1 or task_id > N_TASKS:
        raise ValueError(f"task_id out of range: {task_id} (max {N_TASKS})")
    scenario_id = (task_id - 1) // N_REPS_PER_SCENARIO + 1
    rep         = (task_id - 1) %  N_REPS_PER_SCENARIO + 1
    s           = SCENARIOS[scenario_id - 1]
    return {
        "task_id":     task_id,
        "scenario_id": scenario_id,
        "rep":         rep,
        "n":           s["n"],
        "eta":         s["eta"],
        "alpha_share": s["alpha_share"],
        "name":        s["name"],
    }


def make_seed(scenario_id: int, rep: int) -> int:
    return 1000 * scenario_id + rep


if __name__ == "__main__":
    for s in SCENARIOS:
        print(s)
    print(f"\nTotal scenarios: {N_SCENARIOS}")
    print(f"Total tasks:     {N_TASKS}")

File: Jobs2/without_latent/R/scenarios.R
# =============================================================================
# scenarios.R — JOBS II semi-synthetic, 8 scenarios per Cheng et al. (2022).
# =============================================================================
#
# Scenario grid (from §5.2 of Cheng et al.):
#   n          ∈ {500, 1000}      (sample size to bootstrap)
#   eta        ∈ {1, 10}          (selection-into-mediator strength)
#   alpha_share ∈ {0.10, 0.50}    (target share of "mediated" observations)
# => 2 * 2 * 2 = 8 scenarios, 100 reps each = 800 array tasks per method.
#
# alpha is chosen numerically to hit the target share for each (n, eta).
# The lookup below is filled in by R/calibrate_alpha.R (run once before the
# main pipeline). Default placeholders are reasonable starting guesses; the
# calibrator will overwrite them.
# =============================================================================

SCENARIOS <- expand.grid(
  n           = c(500, 1000),
  eta         = c(1, 10),
  alpha_share = c(0.10, 0.50),
  KEEP.OUT.ATTRS   = FALSE,
  stringsAsFactors = FALSE
)

# These alpha values are PLACEHOLDERS. Run calibrate_alpha.R first to replace
# them with calibrated values. The calibration step writes
# R/alpha_lookup.rds; this file loads it if present.
.alpha_lookup_path <- file.path(
  Sys.getenv("PROJECT_ROOT", unset = "."),
  "R", "alpha_lookup.rds"
)
if (file.exists(.alpha_lookup_path)) {
  ALPHA_LOOKUP <- readRDS(.alpha_lookup_path)
} else {
  # Placeholders (used until calibration runs)
  ALPHA_LOOKUP <- list(
    "n500_eta1_share10"   = -2.3,
    "n500_eta1_share50"   = -0.05,
    "n500_eta10_share10"  = -23.0,
    "n500_eta10_share50"  = -0.5,
    "n1000_eta1_share10"  = -2.3,
    "n1000_eta1_share50"  = -0.05,
    "n1000_eta10_share10" = -23.0,
    "n1000_eta10_share50" = -0.5
  )
}

SCENARIOS$alpha <- mapply(function(n, eta, share) {
  key <- sprintf("n%d_eta%d_share%d", n, eta, share * 100)
  if (!is.null(ALPHA_LOOKUP[[key]])) ALPHA_LOOKUP[[key]] else NA_real_
}, SCENARIOS$n, SCENARIOS$eta, SCENARIOS$alpha_share)

SCENARIOS$scenario_id <- seq_len(nrow(SCENARIOS))
SCENARIOS$name <- with(SCENARIOS,
                      sprintf("n%d_eta%d_share%d",
                              n, eta, alpha_share * 100))

N_REPS_PER_SCENARIO <- 100L   # Cheng et al. used 10; 100 gives tighter CIs

task_id_to_indices <- function(task_id) {
  stopifnot(task_id >= 1,
            task_id <= nrow(SCENARIOS) * N_REPS_PER_SCENARIO)
  scenario_id <- ((task_id - 1L) %/% N_REPS_PER_SCENARIO) + 1L
  rep         <- ((task_id - 1L) %%  N_REPS_PER_SCENARIO) + 1L
  list(
    task_id     = task_id,
    scenario_id = scenario_id,
    rep         = rep,
    n           = SCENARIOS$n[scenario_id],
    eta         = SCENARIOS$eta[scenario_id],
    alpha_share = SCENARIOS$alpha_share[scenario_id],
    alpha       = SCENARIOS$alpha[scenario_id],
    name        = SCENARIOS$name[scenario_id]
  )
}

make_seed <- function(scenario_id, rep) {
  as.integer(1000L * scenario_id + rep)
}
